### PR TITLE
Improve scaling stability

### PR DESCRIPTION
**Pull request by Ladicle**

### CODE REVIEW

File: engine/internal/metrics/metrics.go [hunk 1]
```diff
@@ -2,35 +2,161 @@ package metrics
...
 	defer c.mu.RUnlock()
-	return c.metrics[modelID]
+	wb, ok := c.metrics[modelID]
+	if !ok {
+		return 0
+	}
+	return wb.average(time.Now())
+}
+
+// percision is the granularity of the window bucket.
```
Comment (kenji-cloudnatix) [inline engine/internal/metrics/metrics.go:53]:
```suggestion
// precision is the granularity of the window bucket.
```

File: engine/internal/autoscaler/autoscaler.go [hunk 2]
```diff
@@ -113,16 +114,27 @@ type scaler struct {
...
-func (s *scaler) start(log logr.Logger, stopCh <-chan struct{}, period time.Duration) {
+func (s *scaler) start(log logr.Logger, stopCh <-chan struct{}, initialDelay, period time.Duration) {
 	ctx, cancel := context.WithCancel(context.Background())
 	s.cancel = cancel
 
 	log = log.WithName("scaler").WithValues("modelID", s.modelID)
 	ctx = ctrl.LoggerInto(ctx, log)
 
-	ticker := time.NewTicker(period)
+	runOnce := func() {
```
Comment (kenji-cloudnatix) [inline engine/internal/autoscaler/autoscaler.go:124]: Really personal preference, but how about `scale` for the function name?

File: engine/internal/autoscaler/autoscaler.go [hunk 3]
```diff
@@ -157,21 +163,25 @@ func (s *scaler) scale(ctx context.Context) error {
 		return client.IgnoreNotFound(err)
 	}
 
-	totalObserved := s.metricsClient.Get(s.modelID)
-	observedPerPods := totalObserved / float64(sts.Status.Replicas)
+	// TODO(aya): Stop scaling when there are too many not-ready pods
 
+	metricsVal := s.metricsClient.Get(s.modelID)
+	replicas := math.Min(float64(sts.Status.ReadyReplicas), float64(sts.Status.Replicas))
```
Comment (kenji-cloudnatix) [inline engine/internal/autoscaler/autoscaler.go:169]: might make sense to define our own `min` and `max` and avoid conversion to `float64`.

File: engine/internal/autoscaler/autoscaler.go [hunk 4]
```diff
@@ -182,12 +192,24 @@ func (s *scaler) scale(ctx context.Context) error {
 	} else {
 		s.lastTransitToZero = time.Time{}
 	}
-	return s.scaleTo(ctx, &sts, cappedReplicas)
+
+	specReplicas := float64(ptr.Deref(sts.Spec.Replicas, 0))
```
Comment (kenji-cloudnatix) [inline engine/internal/autoscaler/autoscaler.go:196]: Could you add a brief comment on the logic between line 196-207?

File: engine/internal/config/config.go [hunk 5]
```diff
@@ -191,6 +197,9 @@ type ScalingConfig struct {
 
 	MaxReplicas int `yaml:"maxReplicas"`
 	MinReplicas int `yaml:"minReplicas"`
+
+	MaxScaleUpRate   float64 `yaml:"maxScaleUpRate"`
```
Comment (kenji-cloudnatix) [inline engine/internal/config/config.go:201]: Could you add a brief comment on these fields? Some example case that explains the behavior would be useful.

File: engine/internal/autoscaler/autoscaler.go [hunk 6]
```diff
@@ -182,12 +192,24 @@ func (s *scaler) scale(ctx context.Context) error {
...
+
+	if newReplicas < specReplicas {
+		maxScaleDown := math.Floor(float64(replicas) * s.config.MaxScaleDownRate)
```
Comment (kenji-cloudnatix) [inline engine/internal/autoscaler/autoscaler.go:199]: Checking my understanding. If`MaxScaleDownRate` is 0.2, and the current replica size is 10, we scale down the stateful set to 2 replicas when it is most aggressive. 

When the number of replica becomes 2, `maxScaleDown` is `floor(2*0.2)= 0`. If the min replica is 0, the target replica size becomes 0.

**Review by kenji-cloudnatix — COMMENTED** [review]
Nice! Still looking.

**Comment by Ladicle** [comment]
rebase onto main

**Pull request merged**